js/bam/bamIndex.js

```diff
@@ -221,6 +221,102 @@ function optimizeChunks(chunks, lowest) {
     return mergedChunks;
 }
 
+/**
+ * Returns the number of basepairs that a bin spans.
+ */
+function binSpan(bin) {
+    if (bin === 0) {
+        return Math.pow(2, 29);
+    } else if (1 <= bin && bin <= 8) {
+        return Math.pow(2, 26);
+    } else if (9 <= bin && bin <= 72) {
+        return Math.pow(2, 23);
+    } else if (73 <= bin && bin <= 584) {
+        return Math.pow(2, 20);
+    } else if (585 <= bin && bin <= 4680) {
+        return Math.pow(2, 17);
+    } else {
+        return Math.pow(2, 14);
+    }
+}
+
+/**
+ * Returns the basepairs where a bin's children intersect
+ */
+function bpsWhereChildrenIntersect(bin) {
+    const childSpan = binSpan(bin) / Math.pow(2, 3);
+    const numBinsAtLvl = Math.pow(2, 29) / childSpan;
+    const numIntersectionPoints = numBinsAtLvl - 1;
+
+    let ret = [];
+    let offset = childSpan;
+    for (let i = 0; i < numIntersectionPoints; ++i) {
+        ret.push(offset);
+        offset += childSpan;
+    }
+
+    return ret;
+}
+
+/**
+ * Returns true if the given chunk could overlap with the genomic
+ * range [min, max].
+ */
+function chunkCouldOverlapWithRange(chunk, min, max) {
+    // If the chunk begins and ends with a different `coffset`
+    // (see bam spec) then it must intersect multiple bgzf
+    // blocks. In this case, the uncompressed span cannot be
+    // calculated, so pessimistically assume that the chunk
+    // *could* overlap with the query range.
+    if (chunk.minv.block !== chunk.maxv.block) {
+        return true;
+    }
+
+    // If the chunk is in a "leaf" bin of the rtree, we
+    // pessimistically assume that the chunk *could* overlap with
+    // the query range (it's a leaf bin, so it *probably* does
+    // intersect the query range anyway).
+    if (4681 <= chunk.bin && chunk.bin <= 37448) {
+        return true;
+    }
+
+    // Because all of the data is in one bgzf block (see above),
+    // we can compute the uncompressed span of the chunk
+    const uncompressedSize = chunk.maxv.offset - chunk.minv.offset;
+
+    // Assuming *all* of the data in the span is, in fact, purely
+    // 4-bit basepairs with no quality scores, tag names,
+    // etc. then the theoretical maximum number of base pairs the
+    // chunk may contain can be calculated from its size.
+    const BPS_IN_BYTE = 2;
+    const maxBpsInChunk = BPS_IN_BYTE * uncompressedSize;
+
```

Comment:
I'm not following this logic.  The contents of a bgzf block could consist entirely of alignments at the same genomic location.  Alternatively it could contain alignments that span the entire chromosome.   So I don't understand this 4-bit assumption.  I don't see anyway to determine the maximum # of bps a BGZF block can span without loading and looking at the data.

Reply:
So the logic is that we know how many bytes of uncompressed data that particular chunk represents because we only look at chunks that are within one BGZF block (chunks that span multiple blocks are discarded because compression efficiency can't be assumed).

Lets assume the chunk spans 128 uncompressed BAM bytes. That chunk, in the worst possible case, might contain only BAM basepairs (this isn't necessarily possible, but lets be overly pessimistic). Those base pairs, in the worst case, might be encoded with maximum efficiency (4 bits per base pair). Therefore, in this worst case, the BAM data could only possibly span a range of 256 base-pairs (unless the uncompressed BAM data format contains a mechanism for encoding a larger range than that in 128 bytes). In reality, because the data may contain multiple alignments, quality scores, etc. then the basepair span of the data is going to be lower than this bound.

From that logic, I am assuming that the maximum "span" of that chunk is 256 base pairs. Knowing that span enables this optimization. The optimization algorithm uses the chunk's bin number to infer all possible positions (in base pair space) that the chunk may have spanned when it was being indexed. If the chunk is in a leaf node of the rtree, all we can say is that "the chunk represents an upto 256 bp range within the range of that bin", which isn't useful because earlier steps in the process have already established that the leaf bin is pertinent to the query (`reg2bins`). 

However, if the chunk was assigned  to an internal bin of the tree--and presuming the chunk was assigned to that node *because* it couldn't fit into a smaller bin--then additional information can be inferred. We know, in this particular case, that:

- The chunk is up to some 256 bp range within the bin (same as a leaf)
- And the chunk must have intersected the boundary between two children bin (hence its promotion into an upper bin)
- So, at worst, there are 7 possible ranges in the bin, 512 bps across and centered on the boundaries between direct descendants in the tree, where the chunk *must* have intersected
- And if our query range does not intersect with any of those 7 ranges, then this chunk cannot contain data within our query range

Reply:
Ah, @jrobinso , you're right in that the chunk could contain multiple reads that are aligned to very different parts of the genome, so a counterexample would a chunk containing two reads on opposite ends of the genome, and that chunk being the only chunk in the file (for example).